src/control/lib/spdk/src/nvme_control_common.c

```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
 	return ret;
 }
 
+/* Assign ASCII string as defined by the NVMe spec */
 static int
-copy_ctrlr_data(struct ctrlr_t *cdst, const struct spdk_nvme_ctrlr_data *cdata)
+copy_ascii(char *dst, const void *buf, size_t buf_sz)
 {
-	int	written;
+	const uint8_t	*str = buf;
+	int		 i = 0;
 
-	written = snprintf(cdst->model, sizeof(cdst->model), "%-20.20s",
-			   cdata->mn);
-	if (written >= sizeof(cdst->model))
+	if (buf_sz >= BUFLEN)
 		return -NVMEC_ERR_CHK_SIZE;
 
-	written = snprintf(cdst->serial, sizeof(cdst->serial), "%-20.20s",
-			   cdata->sn);
-	if (written >= sizeof(cdst->serial))
-		return -NVMEC_ERR_CHK_SIZE;
+	/* Trim trailing spaces */
+	while (buf_sz > 0 && str[buf_sz - 1] == ' ')
+		buf_sz--;
 
-	written = snprintf(cdst->fw_rev, sizeof(cdst->fw_rev), "%s", cdata->fr);
-	if (written >= sizeof(cdst->fw_rev))
-		return -NVMEC_ERR_CHK_SIZE;
+	while (buf_sz--) {
+		if (*str >= 0x20 && *str <= 0x7E)
```

Comment:
isprint() could be used if you want check if it's printable characters.

Reply:
as above